web/src/unread_ops.js

```diff
@@ -241,62 +275,100 @@ export function mark_as_unread_from_here(
-                            message_id,
-                            false,
-                            messages_marked_unread_till_now,
-                            narrow,
-                        ),
-                    milliseconds_to_wait,
-                );
-            } else {
-                // TODO: Ideally, this case would communicate the
-                // failure to the user, with some manual retry
-                // offered, since the most likely cause is a 502.
-                blueslip.error("Unexpected error marking messages as unread", {
-                    status: xhr.status,
-                    body: xhr.responseText,
-                });
+            handle_mark_unread_from_here_error(
+                xhr,
+                message_id,
+                messages_marked_unread_till_now,
+                narrow,
+            );
+        },
+    });
+}
+
+function do_mark_unread_by_ids(message_ids_to_update) {
+    channel.post({
+        url: "/json/messages/flags",
+        data: {messages: JSON.stringify(message_ids_to_update), op: "remove", flag: "read"},
+        success() {
+            if (loading_indicator_displayed) {
+                finish_loading(message_ids_to_update.length);
             }
         },
+        error(xhr) {
+            handle_mark_unread_from_here_error(xhr, undefined, 0, undefined, message_ids_to_update);
+        },
     });
 }
 
+function finish_loading(messages_marked_unread_till_now) {
+    // If we were showing a loading indicator, then
+    // display that we finished. For the common case where
+    // the operation succeeds in a single batch, we don't
+    // bother distracting the user with the indication;
+    // the success will be obvious from the UI updating.
+    loading_indicator_displayed = false;
+    ui_report.loading(
+        $t_html(
+            {
+                defaultMessage:
+                    "{N, plural, one {Done! {N} message marked as unread.} other {Done! {N} messages marked as unread.}}",
+            },
+            {N: messages_marked_unread_till_now},
+        ),
+        $("#request-progress-status-banner"),
+        true,
+    );
+}
+
+function handle_mark_unread_from_here_error(
+    xhr,
+    message_id,
+    messages_marked_unread_till_now,
+    narrow,
+    message_ids_to_update = undefined,
+) {
+    if (xhr.readyState === 0) {
+        // client cancelled the request
+    } else if (xhr.responseJSON?.code === "RATE_LIMIT_HIT") {
+        // If we hit the rate limit, just continue without showing any error.
+        const milliseconds_to_wait = 1000 * xhr.responseJSON["retry-after"];
+        if (message_ids_to_update !== undefined) {
+            setTimeout(() => {
+                do_mark_unread_by_ids(message_ids_to_update);
+            }, milliseconds_to_wait);
+        } else {
+            setTimeout(() => {
+                do_mark_unread_by_narrow(
+                    message_id,
+                    false,
```

Comment:
So this isn't new in your code, but I think this `false` here is wrong -- it should be the `include_anchor` value from the request that failed -- which will usually be `false`, but if the very first request failed, will be `true`.

I think the right fix for this probably involves reworking `handle_mark_unread_from_here_error` to accept an object rather than a list of parameters, which would be a lot nicer as we move this file to TypeScript anyway, and then adding `include_anchor` to the parameters passed in here.

Do you want to just do that in a quick follow-up PR? I think that'll save review time, as it'll let me merge this now.

Reply:
Fixed in #30047